sway/commands/gaps.c

```diff
@@ -0,0 +1,153 @@
+			}
+		} else {
+			return cmd_results_new(CMD_INVALID, "gaps", "gaps edge_gaps on|off|toggle");
+		}
+	} else {
+		int amount_idx = 0;
+		char op = '=';
+		char scope = 'a';
+		bool inner = true;
+
+		if (strcmp(argv[amount_idx], "inner") == 0) {
+			amount_idx++;
+			inner = true;
+		} else if (strcmp(argv[amount_idx], "outer") == 0) {
+			amount_idx++;
+			inner = false;
+		}
+
+		if (argc > 2) {
+			if ((error = checkarg(argc, "gaps", EXPECTED_EQUAL_TO, 4))) {
+				return error;
+			}
+
+			if (strcmp(argv[amount_idx], "all") == 0) {
+				amount_idx++;
+				scope = 'a';
+			} else if (strcmp(argv[amount_idx], "workspace") == 0) {
+				amount_idx++;
+				scope = 'w';
+			} else if (strcmp(argv[amount_idx], "current") == 0) {
+				amount_idx++;
+				scope = 'c';
+			}
+
+			if (strcmp(argv[amount_idx], "set") == 0) {
+				amount_idx++;
+				op = '=';
+			} else if (strcmp(argv[amount_idx], "plus") == 0) {
+				amount_idx++;
+				op = '+';
+			} else if (strcmp(argv[amount_idx], "minus") == 0) {
+				amount_idx++;
+				op = '-';
+			}
+		} else {
+			if ((error = checkarg(argc, "gaps", EXPECTED_EQUAL_TO, amount_idx + 1))) {
+				return error;
+			}
+		}
+
+		double val;
+		if (read_double(argv[amount_idx], &val) == 0) {
+			if (amount_idx == 0) {
+				config->gaps_inner = val;
+				config->gaps_outer = val;
+				arrange_root();
+			} else {
+				double total = val;
+				if (op == '-') {
+					total = (inner ? config->gaps_inner : config->gaps_outer) - val;
+					if (total < 0) {
+						total = 0;
+					}
+				} else if (op == '+') {
+					total = (inner ? config->gaps_inner : config->gaps_inner) + val;
+				}
+
+				if (scope == 'a') {
+					if (inner) {
+						config->gaps_inner = total;
+					} else {
+						config->gaps_outer = total;
+					}
+					arrange_root();
+				} else if (scope == 'w') {
+					struct sway_container *workspace =
+					config->handler_context.current_container;
+					if (workspace->type != C_WORKSPACE) {
+						workspace = container_parent(workspace, C_WORKSPACE);
+					}
+					workspace->has_gaps = true;
+					if (inner) {
+						workspace->gaps_inner = total;
+					} else {
+						workspace->gaps_outer = total;
+					}
+					arrange_workspace(workspace);
+				} else if (scope == 'c') {
+					struct sway_container *container =
+					config->handler_context.current_container;
+					container->has_gaps = true;
+					if (inner) {
+						container->gaps_inner = total;
+					} else {
+						container->gaps_outer = total;
+					}
+					arrange_workspace(container_parent(container, C_WORKSPACE));
+				}
+			}
+		} else {
+			if (amount_idx < 1 && argc == 1) {
+				return cmd_results_new(CMD_INVALID, "gaps", "gaps <amount>");
+			} else if (amount_idx < 2 && argc == 2) { // This is the gaps <inner|outer>
+				return cmd_results_new(CMD_INVALID, "gaps", "gaps inner|outer <amount>");
+			} else {
+				return cmd_results_new(CMD_INVALID, "gaps", "gaps inner|outer all|workspace|current set|plus|minus <amount>"); 
+			}
+		}
+	}
```

Comment:
This function overall has some pretty deep indentation, going up to 6 levels. It should be possible to make it a lot shallower by refactoring it and adding some early return statements.

Reply:
I didn't use early returns because I'm personally opposed to them (they're less math-y). I can refactor it.

Reply:
+1 to early returns

Reply:
+1 to early returns

Reply:
I would agree with you more if we were talking about a functional language, or something expression-oriented like Rust. But in a purely imperative language like C, exiting as soon as possible is generally a lot cleaner and can make the logic of the code easier to understand.

Reply:
That makes sense, I've got a haskell/clojure background!